appinventor/components/src/com/google/appinventor/components/runtime/CSVFile.java

```diff
@@ -0,0 +1,280 @@
+        "The results of the reading are stored in the Rows, Columns " +
+        "and ColumnNames properties of the component.")
+    public void ReadFile(String source) {
+        this.sourceFile = source;
+
+        readFromFile(sourceFile);
+    }
+
+    /**
+     * Returns a Future object which holds the CSVFile columns at the point
+     * of invoking the method.
+     *
+     * If reading is in progress, the method blocks until reading is done
+     * before returning the result.
+     *
+     * The method should be called asynchronously to prevent freezing of
+     * the main thread.
+     *
+     * The row size is contained in the method to create default values for the
+     * CharDataModel in case of an absence of columns.
+     *
+     * @param columns  List of columns to retrieve (String object entries expected)
+     * @return  Future object containing YailList of format (rowCount, columns)
+     */
+    public Future<YailList> getColumns(final YailList columns) {
+        // Submit a callable which constructs the results.
+        // The callable is only executed after all the previous
+        // tasks have been completed.
+        return threadRunner.submit(new Callable<YailList>() {
+            @Override
+            public YailList call() {
+                ArrayList<YailList> resultingColumns = new ArrayList<YailList>();
+
+                // Iterate over the specified column names
+                for (int i = 0; i < columns.size(); ++i) {
+                    // Get and add the specified column to the resulting columns list
+                    String columnName = columns.getString(i);
+                    YailList column = getColumn(columnName);
+                    resultingColumns.add(column);
+                }
+
+                // Convert result to a YailList
+                YailList csvColumns = YailList.makeList(resultingColumns);
+
+                // Final result contains the row size as the first element
+                return YailList.makeList(Arrays.asList(rows.size(), csvColumns));
+            }
+        });
+    }
+
+    /**
+     * Gets the specified column's elements as a YailList.
+     *
+     * @param column  name of column
+     * @return YailList of elements in the column
+     */
+    public YailList getColumn(String column) {
+        // Get the index of the column (first row - column names)
+        // 1 is subtracted from the index since YailList indexOf
+        // returns an index that is 1-based.
+        int index = columnNames.indexOf(column) - 1;
+
+        // Column not found
+        if (index < 0) {
+            return new YailList();
+        }
+
+        return (YailList)columns.getObject(index);
+    }
+
+    /**
+     * Instantiates the columns list after CSV parsing to rows
+     * has been processed.
+     */
+    private void constructColumnsFromRows() {
+        // Store columns separately (first row indicates column names)
+        columnNames = (YailList)rows.getObject(0);
```

Comment:
Empty file case unhandled

Reply:
Cases to be tested (in general, both Designer & Android):
- [ ] Empty file
- [ ] Only first row
- [ ] 1 entry only

Reply:
Uneven columns to be fixed in #100